pkg/apis/allocation/v1/gameserverallocation.go

```diff
@@ -82,6 +82,137 @@ type GameServerAllocationSpec struct {
 	MetaPatch MetaPatch `json:"metadata,omitempty"`
 }
 
+// GameServerSelector contains all the filter options for selecting
+// a GameServer for allocation.
+type GameServerSelector struct {
+	metav1.LabelSelector
+	// [Stage:Alpha]
+	// [FeatureFlag:StateAllocationFilter]
+	// +optional
+	GameServerState *agonesv1.GameServerState `json:"gameServerState,omitempty"`
+	// [Stage:Alpha]
+	// [FeatureFlag:PlayerAllocationFilter]
+	// +optional
+	Players *PlayerSelector `json:"players,omitempty"`
+}
+
+// PlayerSelector is the filter options for a GameServer based on player counts
+type PlayerSelector struct {
+	MinAvailable int64 `json:"minAvailable,omitempty"`
+	MaxAvailable int64 `json:"maxAvailable,omitempty"`
+}
+
+// ApplyDefaults applies default values to the PlayerSelector
+func (s *GameServerSelector) ApplyDefaults() {
+	if runtime.FeatureEnabled(runtime.FeatureStateAllocationFilter) {
+		if s.GameServerState == nil {
+			state := agonesv1.GameServerStateReady
+			s.GameServerState = &state
+		}
+	}
+
+	if runtime.FeatureEnabled(runtime.FeaturePlayerAllocationFilter) {
+		if s.Players == nil {
+			s.Players = &PlayerSelector{}
+		}
+	}
+}
+
+// Matches checks to see if a GameServer matches a given GameServerSelector's criteria
+func (s *GameServerSelector) Matches(gs *agonesv1.GameServer) bool {
+
+	// Assume at this point, this has already been run through Validate(), and it can be converted.
```

Comment:
Given that this is an exported function, we are trusting that callers are doing the right thing rather than being defensive and checking the error here. 

WDYT about having an error here cause a panic (instead of returning an error)? Then we make it super obvious that things are broken without polluting the method signature, and we don't have code that ignores errors from functions that it calls.

Reply:
I like this idea a lot! Implementing!